### PR TITLE
add eventType and eventProperties for OutboundLink component, refactor onClick logic too

### PR DESCRIPTION
**Pull request by dfvalero**
I used a similar logic to react-ga for the onClick event (https://github.com/react-ga/react-ga/blob/master/src/components/OutboundLink.js)

I also added eventType and eventProperties, so we will be able to use custom type and props for the event.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,52 +1,52 @@
...
+    if (sameTarget && normalClick) {
+      e.preventDefault();
+      window.amplitude.getInstance().logEvent(amplitudeEventName, amplitudeDetails, () => {
+        window.location.href = href;
+      });
+    } else {
+      window.amplitude.getInstance().logEvent(amplitudeEventName, amplitudeDetails);
+    }
+
+    onClick && onClick(e);
```
Comment (ryanashcraft) [inline src/index.js:23]: Please refactor these function conditions to use `if` condition blocks.

File: src/index.js [hunk 2]
```diff
@@ -1,52 +1,52 @@
...
-  )
+    <a href={href} target={target} onClick={handleClick} onMouseDown={handleMouseDown} {...rest}>
+      {children}
+    </a>
+  );
 }
 
 OutboundLink.propTypes = {
+  eventName: Proptypes.string,
+  eventDetails: Proptypes.any,
```
Comment (ryanashcraft) [inline src/index.js:44]: Can you rename these to eventType and eventProperties to match Amplitude's documentation/terminology for these?

File: src/index.js [hunk 3]
```diff
@@ -1,52 +1,52 @@
...
+        window.location.href = href;
+      });
+    } else {
+      window.amplitude.getInstance().logEvent(amplitudeEventName, amplitudeDetails);
+    }
+
+    onClick && onClick(e);
+  };
+
+  // Adding mouseDown handler to track MIDDLE_CLICK, because MIDDLE_CLICK doesn't reach onClick event
```
Comment (ryanashcraft) [inline src/index.js:26]: Most of this matches react-ga's OutboundLink except for this mouse down handler. Can you explain the inconsistency?

Comment (dfvalero) [reply src/index.js:26]: When you click with the middle/wheel it doesn't reach the onClick event in react (at least in mac). That's why I added this custon onMouse logic, to be able to track middle/wheel clicks.

File: src/index.js (hunk 1 above)
Comment (dfvalero) [reply src/index.js:23]: sure! will do!

File: src/index.js (hunk 2 above)
Comment (dfvalero) [reply src/index.js]: sure!

File: src/index.js (hunk 3 above)
Comment (ryanashcraft) [reply src/index.js:28]: So this is an improvement over react-ga's implementation? Just making sure. And you've tested that it only needs to be applied for middle click and not the other edge-case scenarios (meta/shift clicks)?

Comment (dfvalero) [reply src/index.js:28]: yes, it is an improvement.

I've tested it with middle clicks, shift and ctrl keys in mac. (not windows)

I didn't test it with meta key, I don't know what key it is 😞

Comment (dfvalero) [reply src/index.js:28]: I've tested the command click in mac, and it triggers the logEvent only once.

We can do another thing, we can remove the onMouseDown event. If someone wants to have this logic, he/she will be able to apply this custom logic in their own project.

What do you think?

Comment (ryanashcraft) [reply src/index.js:28]: Yeah, I just am hesitant to add this unless I can be certain that it doesn't ever double-fire on any platform. It'd be great to reference some other popular open source project (like react-ga or another gatsby analytics plugin) that does this in a robust way.

Maybe at this point it'd be better to simplify and de-risk this PR so we can get the other things in? Can always add in this middle click behavior later once you're more confident in it.

Also, this is technically a breaking change, yes? Will need to do another major bump to 4.0.0, don't you think?

**Comment by ryanashcraft** [comment]
Ok, I think I'll go with a minor bump as it's adding new functionality and changing some behavior that shouldn't break any apps, but may change behavior slightly.

I think we're done. Please double check with testing as I don't have time to test, then please rename the PR to represent the current changes and I will use that title for the squashed commit message. Thanks!

**Comment by dfvalero** [comment]
Hi @ryanashcraft I've just tested the code in our project and it works as expected! Thanks for your review :)

**Comment by ryanashcraft** [comment]
@dfvalero Sorry for flip-flopping a bit here. I want to just make sure we don't break anyone currently using this plugin.

Now that you've removed the middle click behavior, how different is this from the original implementation, which was forked from gatsby-plugin-google-analytics? See https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-google-analytics/src/index.js

I can see in a few places where it would have different behaviors, but it seems like those aren't being explicitly called out in this PR. It seems to me the main change here now is just the ability to have a custom event type and event properties per link, no? If so, do you think we could revert back to the original onClick implementation or adapt gatby-plugin-google-analytics' current version, which seems to support custom event types/properties?

**Comment by dfvalero** [comment]
> @dfvalero Sorry for flip-flopping a bit here. I want to just make sure we don't break anyone currently using this plugin.
> 
> Now that you've removed the middle click behavior, how different is this from the original implementation, which was forked from gatsby-plugin-google-analytics? See https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-google-analytics/src/index.js
> 
> I can see in a few places where it would have different behaviors, but it seems like those aren't being explicitly called out in this PR. It seems to me the main change here now is just the ability to have a custom event type and event properties per link, no? If so, do you think we could revert back to the original onClick implementation or adapt gatby-plugin-google-analytics' current version, which seems to support custom event types/properties?

Sure! Totally understandable! I will do a proper PR with the implementation of the eventype and the eventproperties! Thanks for your feedback
